### PR TITLE
Flutter Modules are re-entrant (concurrency)

### PR DESCRIPTION
**Pull request by Larpoux**
This is a major re-design of flutter_sound architecture.
flutter_sound module is now split into : 
- FlautoRecorder
- FlautoPlayer

All modules are re-entrant. It is possible to start several ```FlutterPlayer``` and several ```FlutterRecorder``` at the same time.

Complete backward compatibility is provided by two deprecated modules : 
- FlutterSound.dart
- Flauto.dart

### CODE REVIEW

**Comment by Larpoux** [comment]
Good morning Hyo,

OK. Everything good, now. I replaced `FlautoRecorder` and `FlautoPlayer` by `FlutterSoundRecorder` and `FlutterSoundPlayer`. I replaced `xyz.canardoux.flauto` by `com.dooboolab.flutter_sound`.

I am now ready for a merge to FlutterSound master branch.

Here are the changes : 
```
## 3.1.0

- flutter_sound modules are re-entrant [#250](https://github.com/dooboolab/flutter_sound/issues/250) and [#232](https://github.com/dooboolab/flutter_sound/issues/232)
  - We can open several `FlutterSoundPlayer` at the same time
  - We can open several `FlutterSoundRecorder` at the same time
- Add new API verbs : [#244](https://github.com/dooboolab/flutter_sound/issues/244)
  - flutterSoundHelper.getLastFFmpegReturnCode()
  - flutterSoundHelper.getLastFFmpegCommandOutput()
  - flutterSoundHelper.FFmpegGetMediaInformation() which return info on the given record
  - flutterSoundHelper.duration() which return the number of milli-seconds for the given record
- Add new API verbs : [##242](https://github.com/dooboolab/flutter_sound/issues/242)
  - FlutterSoundRecorder.pauseRecorder()
  - FlutterSoundRecorder.resumeRecorder()
- flutter_sound is now compatible with permission_handler 5.x.x [#259](https://github.com/dooboolab/flutter_sound/issues/259)
- API to control the `audiofocus` [#219](https://github.com/dooboolab/flutter_sound/issues/219)
- API to set the `audio-category` (i.e. duck-others) [#219](https://github.com/dooboolab/flutter_sound/issues/219)
- AndroidX and Android embbeded-V2 support [#203](https://github.com/dooboolab/flutter_sound/issues/203)
- Add a parameter to `startPlayer` to specify a callback when the song is finished [#215](https://github.com/dooboolab/flutter_sound/issues/215)
- License is now LGPL 3.0 instead of MIT
```

File: lib/track_player.dart [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+class TrackPlayerPlugin extends FlautoPlayerPlugin {
+  MethodChannel channel;
+
+  //List<TrackPlayer> trackPlayerSlots = [];
+  TrackPlayerPlugin() {
+    setCallback();
+  }
+
+  void setCallback() {
+    channel = const MethodChannel('xyz.canardoux.track_player');
```
Comment (hyochan) [inline lib/track_player.dart:45]: Any changes in this method channel?

**Comment by hyochan** [comment]
> This is a major re-design of flutter_sound architecture.
> flutter_sound module is now split into :
> 
> * FlautoRecorder
> * FlautoPlayer
> 
> All modules are re-entrant. It is possible to start several `FlutterPlayer` and several `FlutterRecorder` at the same time.
> 
> Complete backward compatibility is provided by two deprecated modules :
> 
> * FlutterSound.dart
> * Flauto.dart

@Larpoux I am currently suffering from file changes you've mentioned. I can't see and `FlutterPlayer` and `FlutterRecorder` in this updates.

**Comment by Larpoux** [comment]
I changed the name. Now the two modules are : 
- FlutterSoundRecorder.dart
- FlutterSoundPlayer.dart

But I did not changed the names of the module in Java and Objective-C. I can do that, now, if you want.
The name of the MethodChannel is not good. I am going to change that, now : 
- com.dooboolab.flutter_sound_track_player
- com.dooboolab.flutter_sound_recorder
- com.dooboolab.flutter_sound_player

**Comment by Larpoux** [comment]
Now, there is 3 channels : 
- One for the Player
- One for the Recorder
- One for the TrackPlayer (flauto extension)

If the App uses several FlutterSoundPlayer, there is several instances of Android Andio Player and several AVAudioPlayer. All those instances are multiplexed on the channel com.dooboolab.flutter_sound_player

If the App uses several FlutterSoundRecorder, there is several instances of Android Audio Recorder and several AVAudioRecorder. All those instances are multiplexed on the channel com.dooboolab.flutter_sound_recorder

**Comment by Larpoux** [comment]
OK Hyo, the 3 channel names are now correct

**Comment by Larpoux** [comment]
Do you want I rename the Java modules and Objective C modules ?

**Comment by Larpoux** [comment]
I really think that *multiplexing* all the flutter_sound instances on the 3 channels is good design.

**Comment by hyochan** [comment]
image

I am still seeing `FlautoPlayer` and not `FlutterSound` tough.

**Comment by hyochan** [comment]
@Larpoux Oh sorry, you've pushed a commit just now~! I'll check that out!

**Comment by hyochan** [comment]
Ok! I've just checked this out and the file names are not changed. Is this what you want?

**Comment by Larpoux** [comment]
Yes, this is what I said.
I will change those name this after-noon. This sill be cleaner.

**Comment by Larpoux** [comment]
Better is Java Modules and ios modules have same name than dart modules

**Comment by Larpoux** [comment]
I will do that after lunch

**Comment by Larpoux** [comment]
Hyo, I fixed a little bug when we try 'Player seek` on iOS.

Seek seems not working with TrackPlayer on Android.
But it works fine with 'FlutterSoundPlayer', and there is no crash. It just does not work with `flauto`.

I think that's no problem for a current release.

**Comment by hyochan** [comment]
> Hyo, I fixed a little bug when we try 'Player seek` on iOS.
> 
> Seek seems not working with TrackPlayer on Android.
> But it works fine with 'FlutterSoundPlayer', and there is no crash. It just does not work with `flauto`.
> 
> I think that's no problem for a current release.

I am on fixing `seeking` in android. Let me do this part~!

**Comment by Larpoux** [comment]
👍 . Great 😸

**Comment by Larpoux** [comment]
Hyo, I renamed the ios modules and Java modules.
I think this is good now.
I am going for my lunch .

**Comment by hyochan** [comment]
CLoses for #261

**Pull request merged**